kms/mackms/mackms.go

```diff
@@ -625,6 +692,70 @@ func getPrivateKey(u *keyAttributes) (*security.SecKeyRef, error) {
 	return security.NewSecKeyRef(key), nil
 }
 
+func getPrivateKeys(u *keyAttributes) ([]*security.SecKeychainItemRef, error) {
+	dict := cf.Dictionary{
+		security.KSecClass:            security.KSecClassKey,
+		security.KSecAttrKeyClass:     security.KSecAttrKeyClassPrivate,
+		security.KSecReturnAttributes: cf.True, // return keychain attributes, i.e. tag and label
+		security.KSecMatchLimit:       security.KSecMatchLimitAll,
+	}
+
+	if u.tag != "" {
+		cfTag, err := cf.NewData([]byte(u.tag))
+		if err != nil {
+			return nil, err
+		}
+		defer cfTag.Release()
+		dict[security.KSecAttrApplicationTag] = cfTag
+	}
+	if u.label != "" {
+		cfLabel, err := cf.NewString(u.label)
+		if err != nil {
+			return nil, err
+		}
+		defer cfLabel.Release()
+		dict[security.KSecAttrLabel] = cfLabel
+	}
+	if len(u.hash) > 0 {
+		cfHash, err := cf.NewData(u.hash)
+		if err != nil {
+			return nil, err
+		}
+		defer cfHash.Release()
+		dict[security.KSecAttrApplicationLabel] = cfHash
+	}
+
+	// construct the query
+	query, err := cf.NewDictionary(dict)
+	if err != nil {
+		return nil, err
+	}
+	defer query.Release()
+
+	// perform the query
+	var result cf.TypeRef
+	err = security.SecItemCopyMatching(query, &result)
+	if err != nil {
+		if errors.Is(err, security.ErrNotFound) {
+			return []*security.SecKeychainItemRef{}, nil
+		}
+		return nil, fmt.Errorf("macOS SecItemCopyMatching failed: %w", err)
+	}
+
+	array := cf.NewArrayRef(result)
+	defer array.Release()
+
+	keys := make([]*security.SecKeychainItemRef, array.Len())
+	for i := 0; i < array.Len(); i++ {
+		item := array.Get(i)
+		key := security.NewSecKeychainItemRef(item)
+		key.Retain() // retain the key, so that it's not released early
```

Comment:
I suppose this gets released early because of the `array.Release()`, right?

Reply:
Not retaining it here will result in an error on getting the first property from the dictionary (note that it fails on `GetSecAttrTokenID`, after some reordering):

```
runtime.cgocall(0x100a5e870, 0x1400012da28)
	/Users/herman/sdk/go1.21.0/src/runtime/cgocall.go:157 +0x44 fp=0x1400012d9f0 sp=0x1400012d9b0 pc=0x100709fc4
go.step.sm/crypto/internal/darwin/security._Cfunc_CFDictionaryGetValue(0x12821a9e0, 0x1e5d71890)
	_cgo_gotypes.go:363 +0x34 fp=0x1400012da20 sp=0x1400012d9f0 pc=0x10096dae4
go.step.sm/crypto/internal/darwin/security.GetSecAttrTokenID.func1(0xffffffffffffffff?)
```